backend/apid/actions/delete_events_by_entity.go

```diff
@@ -43,5 +42,14 @@ func (d EntityDeleter) Delete(req *http.Request) (interface{}, error) {
 		}
 	}
 
-	return nil, d.EntityStore.DeleteEntityByName(req.Context(), entityName)
+	result, err := d.Store.GetEntityByName(req.Context(), entityName)
+	if err != nil {
+		return nil, NewError(InternalErr, err)
+	}
+
+	if result == nil {
+		return nil, NewErrorf(NotFound)
+	}
+
+	return nil, d.Store.DeleteEntityByName(req.Context(), entityName)
```

Comment:
I sympathize that it's often confusing to issue a successful delete command, when no such entity existed. However, the original code actually has correct behaviour.

We should not check for existence before issuing the delete, and we should not return an error for requests that issue a delete for a non-existent resource. Doing so would violate the idempotency property of delete requests, and it would also be a breaking change for our HTTP API.

See https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html section 9.1.2 for more information about idempotency.

If other parts of the server exhibit this property, they should also be changed to not return an error, as they would not be conforming to how HTTP delete is supposed to work.

If we want sensuctl to have this behaviour, we could issue a GET followed by a DELETE, if the GET succeeds, in sensuctl.

Reply:
I'm a little confused, as we use similar logic for other delete endpoints:
https://github.com/sensu/sensu-go/blob/dae00ac999bb340132af11319a4695649801a4d8/backend/apid/actions/events.go#L71-L85

Additionally, [this test](https://github.com/sensu/sensu-go-qa-crucible/blob/e3fba2059f6a5fb9a5494b21c4816daafafdc23d/spec/backend/entities_spec.rb#L31-L42) (and for all other resources) indicates that we always have returned an error when attempting to delete a resource that does not exist. The fact that this test has recently started to fail indicates that the original behavior must have changed in order to cause this bug.

If idempotency is the issue here, then I would suggest applying that same idea consistently across all resources. For example, the generic client and API functions return errors for the same reason (and entities are a snowflake).
```
$ sensuctl check delete test-check --skip-confirm
Error: not found
```

Reply:
I believe that is incorrect behaviour for an HTTP server.

I can't explain why the test has recently started to fail however. That should definitely be investigated!

Reply:
Edit: I'm wrong about the response code. Idempotency is about effects on the state of the server, not the response code.